
```ruby
begin
  raise "boom" # => raise RuntimeError, "boom"
end
#  boom (RuntimeError)

begin
  do_something
rescue NameError
end

raise ArgumentError, "two arguments are needed" # catch only ArgumentError error
raise LoadError, "file not found" # catch only LoadError error

begin
rescue # => rescue StandardError
end

begin #multi error catch block
rescue ArgumentError,NameError
end

rescue StandardError => e
  p e.class.name
  p e.message
  p e.backtrace

class NotAuthorizedError < StandardError; end; # => custom error
NotAuthorizedError.new("You are not authorized to edit record");
# feeling fancy???
class NotAuthorizedError < StandardError
  attr_reader :account_id
  def initialize(message, account_id)
    super(message) #invoke the constructor of parent to set the message
    @account_id = account_id
  end
end
raise NotAuthorizedError.new("Not authorized", 171)


do_something rescue nil #  same as the below code
begin
  do_something
rescue
  nil
end

# ApplicationController
class ApplicationController < ActionController::Base
  rescue_from ActiveRecord::RecordNotFound, with: :render_404

  def render_404
  end
end

class ArticlesController < ApplicationController
  def show
    @article = Article.find(params[:id])
  end
end

def load_quiz
  @quiz = current_user.quizzes.find(params[:id])
rescue ActiveRecord::RecordNotFound
  format.json { render status: :not_found, json: { error: "Quiz not found"}}
end


```

Exception
  NoMemoryError
  ScriptError
    LoadError
    NotImplementedError
    SyntaxError
  SecurityError
  SignalException
    Interrupt
  StandardError
    ArgumentError
      UncaughtThrowError
    EncodingError
    FiberError
    IOError
      EOFError
    IndexError
      KeyError
      StopIteration
    LocalJumpError
    NameError
      NoMethodError
    RangeError
      FloatDomainError
    RegexpError
    RuntimeError
    SystemCallError
      Errno::*
    ThreadError
    TypeError
    ZeroDivisionError
  SystemExit
  SystemStackError
